game.js

```diff
@@ -547,69 +547,60 @@ module.exports = function createGame(options) {
+            throw new GameException('Dead players cannot challenge');
         }
-        else if (command.stateId != state.stateId) {
-            throw new GameException('Stale state (' + command.stateId + '!=' + state.stateId + ')');
+    };
+    const playersTurn = (command, playerIdx) => {
+        if (state.state.playerIdx != playerIdx) {
+            throw new GameException('Not your turn');
         }
-        else if (command.command == 'start') {
-            if (playerState.isReady !== true) {
-                throw new GameException('You cannot start the game');
-            }
-            start(command.gameType);
-
-        } else if (command.command == 'ready') {
-            if (state.state.name != stateNames.WAITING_FOR_PLAYERS) {
-                throw new GameException('Incorrect state');
-            }
-            playerReady(playerIdx);
-
-        } else if (command.command == 'add-ai') {
-            if (state.state.name != stateNames.WAITING_FOR_PLAYERS) {
-                throw new GameException('Incorrect state');
-            }
-            if (playerState.isReady !== true) {
-                throw new GameException('You cannot add AI players');
-            }
-            addAiPlayer();
-
-        } else if (command.command == 'remove-ai') {
-            if (state.state.name != stateNames.WAITING_FOR_PLAYERS) {
-                throw new GameException('Incorrect state');
-            }
-            if (playerState.isReady !== true) {
-                throw new GameException('You cannot remove AI players');
-            }
-            removeAiPlayer();
-
-        } else if (command.command == 'play-action') {
-            if (state.state.name != stateNames.START_OF_TURN) {
-                throw new GameException('Incorrect state');
-            }
-            if (state.state.playerIdx != playerIdx) {
-                throw new GameException('Not your turn');
-            }
-            action = actions[command.action];
+    };
+    const invalidPlayer = (stateField) => (command, playerIdx) => {
+        if (playerIdx == state.state[stateField]) {
+            throw new GameException(`Player cannot ${command.command}`);
+        }
+    };
+    // the mapping of command -> behavior, or command,state -> behavior
+    const behaviors = {};
+    behaviors['start'] = [
+        playerIsReady,
+        (command, playerIdx) => start(command.gameType)
+    ];
+    behaviors['leave'] = [
+        (command, playerIdx) => playerLeft(playerIdx)
+    ];
+    behaviors[['ready', stateNames.WAITING_FOR_PLAYERS]] = [(command, playerIdx) => {
+        playerReady(playerIdx);
+    }];
+    behaviors[['add-ai', stateNames.WAITING_FOR_PLAYERS]] = [
+        playerIsReady,
+        addAiPlayer
+    ];
+    behaviors[['remove-ai', stateNames.WAITING_FOR_PLAYERS]] = [
+        playerIsReady,
+        removeAiPlayer
+    ];
+    behaviors[['play-action', stateNames.START_OF_TURN]] = [
+        playersTurn,
+        (command, playerIdx) => {
```

Comment:
I like how the behaviours are all very readable at a glance now. Maybe factor all of them out into named functions?